File: src/com/sywood/starbucks/andi/Java/mockccc15j5.java
package com.sywood.starbucks.andi.Java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * @author Andi Gu
 */
public class mockccc15j5 {
    private static final Comparator<int[]> h_compare = (o1, o2) -> {
        int result = Integer.compare(o1[0], o2[0]);
        if (result == 0) {
            result = Integer.compare(o1[1], o2[1]);
        }
        return result;
    };
    private static final Comparator<int[]> v_compare = (o1, o2) -> {
        int result = Integer.compare(o1[1], o2[1]);
        if (result == 0) {
            result = Integer.compare(o1[0], o2[0]);
        }
        return result;
    };
    private static int n_buildings;
    private static List<int[]> horizontal;
    private static List<int[]> vertical;

    private static int bisect(int[] x, boolean isHorizontal) {
        Comparator<int[]> comparator = isHorizontal ? h_compare : v_compare;
        List<int[]> list = isHorizontal ? horizontal : vertical;
        int low = 0, high = n_buildings;
        while (low < high) {
            int mid = (low + high) / 2;
            if (comparator.compare(x, list.get(mid)) < 0) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }

    private static int bisect_end(int[] x, boolean isHorizontal) {
        Comparator<int[]> comparator = isHorizontal ? h_compare : v_compare;
        List<int[]> list = isHorizontal ? horizontal : vertical;
        int low = 0, high = n_buildings;
        while (low < high) {
            int mid = (low + high) / 2;
            if (comparator.compare(x, list.get(mid)) <= 0) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }

    public static void main(String[] tokens) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        n_buildings = Integer.parseInt(reader.readLine().split(" ")[0]);
        horizontal = new ArrayList<>();
        vertical = new ArrayList<>();
        for (int i = 0; i < n_buildings; i++) {
            tokens = reader.readLine().split(" ");
            int[] coordinate = new int[]{Integer.parseInt(tokens[0]), Integer.parseInt(tokens[1])};
            horizontal.add(coordinate);
            vertical.add(coordinate);
        }
        horizontal.sort(h_compare);
        vertical.sort(v_compare);

        int n_turnings = Integer.parseInt(reader.readLine().split(" ")[0]);
        int[][] turnings = new int[n_turnings][2];
        for (int i = 0; i < n_turnings; i++) {
            tokens = reader.readLine().split(" ");
            turnings[i] = new int[]{Integer.parseInt(tokens[0]), Integer.parseInt(tokens[1])};
        }


        long result = 0;
        for (int i = 0; i < n_turnings - 1; i++) {
            if (turnings[i][0] == turnings[i + 1][0]) {
                if (turnings[i][1] > turnings[i + 1][1]) {
                    result += bisect(turnings[i], true) - bisect_end(turnings[i + 1], true);
                } else {
                    result += bisect(turnings[i + 1], true) - bisect_end(turnings[i], true);
                }
            } else {
                if (turnings[i][0] > turnings[i + 1][0]) {
                    result += bisect(turnings[i], false) - bisect_end(turnings[i + 1], false);
                } else {
                    result += bisect(turnings[i + 1], false) - bisect_end(turnings[i], false);
                }
            }
        }
        System.out.println(result);
    }
}

File: src/com/sywood/starbucks/andi/Python/mockccc15j5.py
import sys


def bisect_h(x):
    lo, hi = 0, n_buildings
    while lo < hi:
        mid = (lo + hi) // 2
        if x < horizontal[mid]:
            hi = mid
        else:
            lo = mid + 1
    return lo


def bisect_v(x):
    x = x[1], x[0]
    lo, hi = 0, n_buildings
    while lo < hi:
        mid = (lo + hi) // 2
        if x < vertical[mid]:
            hi = mid
        else:
            lo = mid + 1
    return lo


input = sys.stdin.readline
n_buildings = int(input())
buildings = {tuple(int(i) for i in input().split()) for j in range(n_buildings)}

n_turnings = int(input())
turnings = [tuple(int(i) for i in input().split()) for j in range(n_turnings)]
horizontal = sorted(buildings)
vertical = sorted((i[1], i[0]) for i in buildings)
result = 0
dp_h = {}
dp_v = {}
for i in range(n_turnings - 1):
    if turnings[i][0] == turnings[i + 1][0]:
        if turnings[i] not in dp_h:
            dp_h[turnings[i]] = bisect_h(turnings[i])
        if turnings[i + 1] not in dp_h:
            dp_h[turnings[i + 1]] = bisect_h(turnings[i + 1])
        if turnings[i][1] > turnings[i + 1][1]:
            result += dp_h[turnings[i]] - dp_h[turnings[i + 1]] + (turnings[i + 1] in buildings)
        else:
            result += dp_h[turnings[i + 1]] - dp_h[turnings[i]] + (turnings[i] in buildings)
    else:
        if turnings[i] not in dp_v:
            dp_v[turnings[i]] = bisect_v(turnings[i])
        if turnings[i + 1] not in dp_v:
            dp_v[turnings[i + 1]] = bisect_v(turnings[i + 1])

        if turnings[i][0] > turnings[i + 1][0]:
            result += dp_v[turnings[i]] - dp_v[turnings[i + 1]] + (turnings[i + 1] in buildings)
        else:
            result += dp_v[turnings[i + 1]] - dp_v[turnings[i]] + (turnings[i] in buildings)

print(result)